domain/secret/service/watcher.go

```diff
@@ -122,108 +122,15 @@ func (s *WatchableService) WatchRemoteConsumedSecretsChanges(ctx context.Context
-	)
-	if err != nil {
-		return nil, errors.Capture(err)
-	}
-	processChanges := func(ctx context.Context, revisionUUIDs ...string) ([]watcher.SecretTriggerChange, error) {
-		result, err := s.secretState.GetSecretsRevisionExpiryChanges(ctx, appOwners, unitOwners, revisionUUIDs...)
-		if err != nil {
-			return nil, errors.Capture(err)
-		}
-		changes := make([]watcher.SecretTriggerChange, len(result))
-		for i, r := range result {
-			changes[i] = watcher.SecretTriggerChange{
-				URI:             r.URI,
-				Revision:        r.Revision,
-				NextTriggerTime: r.NextTriggerTime,
-			}
-		}
-		return changes, nil
-	}
-	return newSecretStringWatcher(w, s.logger, processChanges)
-}
-
-type obsoleteWatcher struct {
-	catacomb       catacomb.Catacomb
-	logger         logger.Logger
-	state          State
-	watcherFactory WatcherFactory
-
-	appOwners  domainsecret.ApplicationOwners
-	unitOwners domainsecret.UnitOwners
-
-	// sourceWatcher watches for both `All` secret events and obsolete revision `Changed` changes.
-	// So the events are secret URIs and obsolete revision UUIDs.
-	sourceWatcher watcher.StringsWatcher
-
-	// knownSecretURIs is a set of currently owned secret URIs.
-	// Tracking this set allows us to identify if a deletion event corresponds to a previously owned secret.
-	// When a deletion event is received, the secret data is no longer available in the database,
-	// so we cannot query the database to determine if the secret was previously owned.
-	knownSecretURIs set.Strings
-
-	out chan []string
-}
-
-func newObsoleteWatcher(
-	state State,
-	watcherFactory WatcherFactory,
-	logger logger.Logger,
-	owners ...CharmSecretOwner,
-) (*obsoleteWatcher, error) {
-	if len(owners) == 0 {
-		return nil, errors.Errorf("at least one owner must be provided for the obsolete secret watcher")
-	}
-
 	appOwners, unitOwners := splitCharmSecretOwners(owners...)
-	w := &obsoleteWatcher{
-		logger:          logger,
-		state:           state,
-		watcherFactory:  watcherFactory,
-		appOwners:       appOwners,
-		unitOwners:      unitOwners,
-		knownSecretURIs: set.NewStrings(),
-		out:             make(chan []string),
-	}
 
-	if err := w.init(); err != nil {
-		return nil, errors.Errorf("initialising the obsolete secret watcher: %w", err)
-	}
-
-	err := catacomb.Invoke(catacomb.Plan{
-		Name: "obsolete-secret-watcher",
-		Site: &w.catacomb,
-		Work: w.loop,
-		Init: []worker.Worker{w.sourceWatcher},
-	})
-	return w, errors.Capture(err)
-}
-
-func (w *obsoleteWatcher) init() error {
-	tableSecrets, querySecrets := w.state.InitialWatchStatementForOwnedSecrets(w.appOwners, w.unitOwners)
-	tableObsoleteRevisions, queryObsoleteRevisions := w.state.InitialWatchStatementForObsoleteRevision(
-		w.appOwners, w.unitOwners,
+	tableSecrets, querySecrets := s.secretState.InitialWatchStatementForOwnedSecrets(appOwners, unitOwners)
+	tableObsoleteRevisions, queryObsoleteRevisions := s.secretState.InitialWatchStatementForObsoleteRevision(
+		appOwners, unitOwners,
```

Comment:
```suggestion
	tableSecrets, tableObsoleteRevisions, initialQuery := s.secretState.InitialQueryWatchObsolete(appOwners, unitOwners)
```
Where in state
```
func (st State) InitialQueryWatchObsolete(
 	appOwners domainsecret.ApplicationOwners,
 	unitOwners domainsecret.UnitOwners,
 ) (string, string, eventsource.NamespaceQuery , error) {}
```
pushing the inline func `initialQuery` below into state.

It's a pattern used in other domains. With the added benefit of making the `WatchObsolete` method shorter and easier to understand.

Reply:
Since we have separate logic to handle changes from the two tables in the service layer, it’s better to keep them separate and `explicit`. Merging them into one would add unnecessary complexity and make the code harder to understand.